### PR TITLE
(PDB-3639) Optimize existing-fact-values query

### PR DESCRIPTION
**Pull request by mullr**
This was previously done with the text version of the value_hash column; it's
stored as bytea, so it required a conversion. Change the query to compare in the
bytea domain.

### CODE REVIEW

**Comment by mullr** [comment]
Fact storage benchmark results
Before: 
    medium: 123525.899966 ms
After:
    medium: 87886.204174 ms

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 1]
```diff
@@ -991,31 +991,31 @@
...
-               (sutils/array-to-param "text" String values))]
-          [value id])))
-
-(defn realize-records!
+  [vhashes]
+  (let [vhashes-param (sutils/array-to-param "bytea" PGobject
+                                             (map sutils/munge-hash-for-storage vhashes))
+        rows (query-to-vec "WITH new_hashes AS (SELECT unnest(?) as h)
+                              SELECT id, encode(value_hash::bytea, 'hex') value_hash FROM fact_values
+                              INNER JOIN new_hashes ON new_hashes.h = value_hash"
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/storage.clj]: I think we want value_hash on the RHS here to use the index.

Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/storage.clj]: LHS rather

**Comment by wkalt** [comment]
what's the exact test you're using?

File: src/puppetlabs/puppetdb/scf/storage.clj (hunk 1 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/storage.clj]: I think it's true that the index won't get used if value_hash is on the RHS, but in practice it doesn't seem to be getting used in either case on the limited set I'm looking at. May be worth digging around to confirm that we can use it.

Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/storage.clj]: poking a little more it looks like the order shouldn't matter, so I retract everything except that we should try and get confirmation that it'll use an index sometimes.

**Comment by mullr** [comment]
@wkalt these timings are https://github.com/puppetlabs/puppetdb/pull/2355, using the 'medium' profile.

File: src/puppetlabs/puppetdb/scf/storage.clj (hunk 1 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/scf/storage.clj]:
```
                                                                  QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.93..839.51 rows=100 width=40) (actual time=0.033..0.049 rows=3 loops=1)
   CTE new_hashes
     ->  Result  (cost=0.00..0.51 rows=100 width=32) (actual time=0.006..0.008 rows=3 loops=1)
   ->  CTE Scan on new_hashes  (cost=0.00..2.00 rows=100 width=32) (actual time=0.007..0.010 rows=3 loops=1)
   ->  Index Scan using fact_values_value_hash_key on fact_values  (cost=0.42..8.36 rows=1 width=29) (actual time=0.008..0.008 rows=1 loops=3)
         Index Cond: (value_hash = new_hashes.h)
 Planning time: 0.141 ms
 Execution time: 0.075 ms
(8 rows)
```

**Comment by mullr** [comment]
@puppetlabs-jenkins retest this please

**Pull request merged**